File: frontend/station-tracker-geolocation/ios/Plugin/StationTrackerGeolocationPlugin.swift
import Foundation
import Capacitor
import CoreLocation

@objc(StationTrackerGeolocationPlugin)
public class StationTrackerGeolocationPlugin: CAPPlugin {
    private let locationManager = LocationManager()

    override public func load() {
        locationManager.plugin = self
    }

    @objc func configure(_ call: CAPPluginCall) {
        guard let apiUrl = call.getString("apiUrl") else {
            call.reject("apiUrl is required")
            return
        }
        guard let apiKey = call.getString("apiKey") else {
            call.reject("apiKey is required")
            return
        }
        guard let userId = call.getString("userId") else {
            call.reject("userId is required")
            return
        }
        let endpoint = call.getString("endpoint") ?? "/dev/location"

        // Save config to UserDefaults
        let defaults = UserDefaults.standard
        defaults.set(apiUrl, forKey: "StationTracker.apiUrl")
        defaults.set(apiKey, forKey: "StationTracker.apiKey")
        defaults.set(userId, forKey: "StationTracker.userId")
        defaults.set(endpoint, forKey: "StationTracker.endpoint")

        call.resolve(["success": true])
    }

    @objc func startTracking(_ call: CAPPluginCall) {
        let distanceFilter = call.getInt("distanceFilter") ?? 10
        let notificationTitle = call.getString("notificationTitle") ?? "駅記録アプリ"
        let notificationText = call.getString("notificationText") ?? "バックグラウンドで位置を追跡中"
        let stationaryRadius = call.getInt("stationaryRadius") ?? 50
        let stationaryTime = call.getInt("stationaryTime") ?? 300

        // Save tracking options
        let defaults = UserDefaults.standard
        defaults.set(distanceFilter, forKey: "StationTracker.distanceFilter")
        defaults.set(notificationTitle, forKey: "StationTracker.notificationTitle")
        defaults.set(notificationText, forKey: "StationTracker.notificationText")
        defaults.set(stationaryRadius, forKey: "StationTracker.stationaryRadius")
        defaults.set(stationaryTime, forKey: "StationTracker.stationaryTime")
        defaults.set(true, forKey: "StationTracker.isTracking")

        locationManager.startTracking()

        call.resolve(["success": true])
    }

    @objc func stopTracking(_ call: CAPPluginCall) {
        let defaults = UserDefaults.standard
        defaults.set(false, forKey: "StationTracker.isTracking")

        locationManager.stopTracking()

        call.resolve(["success": true])
    }

    @objc func getTrackingStatus(_ call: CAPPluginCall) {
        let defaults = UserDefaults.standard
        let isTracking = defaults.bool(forKey: "StationTracker.isTracking")

        call.resolve(["isTracking": isTracking])
    }

    @objc func getQueueSize(_ call: CAPPluginCall) {
        let size = LocationDatabase.shared.getPendingCount()
        call.resolve(["size": size])
    }

    @objc func forceSendQueue(_ call: CAPPluginCall) {
        locationManager.forceSendQueue()
        call.resolve(["sent": 0, "failed": 0])
    }

    func notifyLocationUpdate(latitude: Double, longitude: Double, accuracy: Double, timestamp: Date) {
        notifyListeners("locationUpdate", data: [
            "latitude": latitude,
            "longitude": longitude,
            "accuracy": accuracy,
            "timestamp": ISO8601DateFormatter().string(from: timestamp)
        ])
    }

    func notifyVisitDetected(latitude: Double, longitude: Double, arrivedAt: Date) {
        notifyListeners("visitDetected", data: [
            "latitude": latitude,
            "longitude": longitude,
            "arrivedAt": ISO8601DateFormatter().string(from: arrivedAt)
        ])
    }

    func notifyError(message: String) {
        notifyListeners("error", data: ["message": message])
    }
}

File: frontend/station-tracker-geolocation/android/src/main/java/com/stationtracker/geolocation/StationTrackerGeolocationPlugin.kt
package com.stationtracker.geolocation

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import androidx.core.content.ContextCompat
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.getcapacitor.annotation.PermissionCallback

@CapacitorPlugin(
    name = "StationTrackerGeolocation",
    permissions = [
        Permission(strings = [Manifest.permission.ACCESS_FINE_LOCATION], alias = "location"),
        Permission(strings = [Manifest.permission.ACCESS_COARSE_LOCATION], alias = "coarseLocation"),
        Permission(strings = [Manifest.permission.ACCESS_BACKGROUND_LOCATION], alias = "backgroundLocation")
    ]
)
class StationTrackerGeolocationPlugin : Plugin() {

    private lateinit var locationDatabase: LocationDatabase

    override fun load() {
        locationDatabase = LocationDatabase(context)
    }

    @PluginMethod
    fun configure(call: PluginCall) {
        val apiUrl = call.getString("apiUrl") ?: run {
            call.reject("apiUrl is required")
            return
        }
        val apiKey = call.getString("apiKey") ?: run {
            call.reject("apiKey is required")
            return
        }
        val userId = call.getString("userId") ?: run {
            call.reject("userId is required")
            return
        }
        val endpoint = call.getString("endpoint", "/dev/location")

        // Save config to SharedPreferences
        val prefs = context.getSharedPreferences("StationTrackerGeo", android.content.Context.MODE_PRIVATE)
        prefs.edit().apply {
            putString("apiUrl", apiUrl)
            putString("apiKey", apiKey)
            putString("userId", userId)
            putString("endpoint", endpoint)
            apply()
        }

        val ret = JSObject()
        ret.put("success", true)
        call.resolve(ret)
    }

    @PluginMethod
    fun startTracking(call: PluginCall) {
        if (!hasRequiredPermissions()) {
            requestAllPermissions(call, "startTrackingPerms")
            return
        }

        val distanceFilter = call.getInt("distanceFilter", 10)
        val notificationTitle = call.getString("notificationTitle", "駅記録アプリ")
        val notificationText = call.getString("notificationText", "バックグラウンドで位置を追跡中")
        val stationaryRadius = call.getInt("stationaryRadius", 50)
        val stationaryTime = call.getInt("stationaryTime", 300)

        // Save tracking options to SharedPreferences
        val prefs = context.getSharedPreferences("StationTrackerGeo", android.content.Context.MODE_PRIVATE)
        prefs.edit().apply {
            putInt("distanceFilter", distanceFilter)
            putString("notificationTitle", notificationTitle)
            putString("notificationText", notificationText)
            putInt("stationaryRadius", stationaryRadius)
            putInt("stationaryTime", stationaryTime)
            putBoolean("isTracking", true)
            apply()
        }

        // Start foreground service
        val serviceIntent = Intent(context, LocationTrackingService::class.java)
        ContextCompat.startForegroundService(context, serviceIntent)

        // Schedule daily maintenance
        WorkManagerHelper.scheduleDailyMaintenance(context)

        val ret = JSObject()
        ret.put("success", true)
        call.resolve(ret)
    }

    @PermissionCallback
    private fun startTrackingPerms(call: PluginCall) {
        if (hasRequiredPermissions()) {
            startTracking(call)
        } else {
            call.reject("Location permission denied")
        }
    }

    @PluginMethod
    fun stopTracking(call: PluginCall) {
        val prefs = context.getSharedPreferences("StationTrackerGeo", android.content.Context.MODE_PRIVATE)
        prefs.edit().putBoolean("isTracking", false).apply()

        val serviceIntent = Intent(context, LocationTrackingService::class.java)
        context.stopService(serviceIntent)

        // Cancel daily maintenance
        WorkManagerHelper.cancelDailyMaintenance(context)

        val ret = JSObject()
        ret.put("success", true)
        call.resolve(ret)
    }

    @PluginMethod
    fun getTrackingStatus(call: PluginCall) {
        val prefs = context.getSharedPreferences("StationTrackerGeo", android.content.Context.MODE_PRIVATE)
        val isTracking = prefs.getBoolean("isTracking", false)

        val ret = JSObject()
        ret.put("isTracking", isTracking)
        call.resolve(ret)
    }

    @PluginMethod
    fun getQueueSize(call: PluginCall) {
        val size = locationDatabase.getPendingCount()

        val ret = JSObject()
        ret.put("size", size)
        call.resolve(ret)
    }

    @PluginMethod
    fun forceSendQueue(call: PluginCall) {
        // Trigger immediate send via service
        val serviceIntent = Intent(context, LocationTrackingService::class.java).apply {
            action = LocationTrackingService.ACTION_FORCE_SEND
        }
        context.startService(serviceIntent)

        val ret = JSObject()
        ret.put("sent", 0)  // Will be updated via callback
        ret.put("failed", 0)
        call.resolve(ret)
    }

    private fun hasRequiredPermissions(): Boolean {
        return ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
    }

    fun notifyLocationUpdate(latitude: Double, longitude: Double, accuracy: Float, timestamp: Long) {
        val ret = JSObject().apply {
            put("latitude", latitude)
            put("longitude", longitude)
            put("accuracy", accuracy)
            put("timestamp", timestamp)
        }
        notifyListeners("locationUpdate", ret)
    }

    fun notifyVisitDetected(latitude: Double, longitude: Double, arrivedAt: Long) {
        val ret = JSObject().apply {
            put("latitude", latitude)
            put("longitude", longitude)
            put("arrivedAt", arrivedAt)
        }
        notifyListeners("visitDetected", ret)
    }

    fun notifyError(message: String) {
        val ret = JSObject().apply {
            put("message", message)
        }
        notifyListeners("error", ret)
    }
}
